rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+primarily as a traditional logging library to this point, leveling spans and
+having those levels coupled to logging statements can feel a bit weird. To keep
+things relatively simple, **we propose that the spans discussed here be given
+a default level of `INFO`**. This will make them available without the
+additional noise that comes with increased logging levels, and should not cause
+a significant performance impact given the relatively broad spans we're
+recommending (i.e. no high-throughput, per-event spans).
+
+One final concern when instrumenting with spans is the length of the span's
+lifetime. In early experiments, we ran into issues where extremely long-lived
+spans (e.g. covering the full runtime of a component) are opened and collect
+data continuously, never flushing because the span does not end. This lead to
+extreme amount of memory growth. The problem was greatly exaggerated by enabling
+`TRACE`-level logging, which resulted in an enormous number of log events
+collecting within the span. To address this, **we proposing pushing the
+outermost spans down from the lifetime of a component to an "iteration" of
+a component**. This will depend greatly on the specifics of each component, but
+the file source above can serve as an example. We also propose keeping these
+spans at the `INFO` level, where our logging is actually pretty quiet and there
+is far less potential for accumulating memory even if spans last longer than
+intended.
+
+In summary:
+
+- Spans should be component-focused, breaking down their operations by stages with at least two levels of hierarchy but up to three levels for components with more complex internals:
+    1. Single "iteration" of operation, on the order of seconds, as a baseline
+    2. High level stages of operation within an iteration, such as reading,
+       processing, sending
+    3. Interesting sub-stages of each operation, with a focus on units of work
+       with significant off-CPU time
+- Spans should be instantiated at the `INFO` log level
+
+#### Forwarding
+
+For the mechanics of actually sending span data to Datadog, we propose relying
+on a couple of existing libraries to handle virtually all of the heavy lifting.
+It's possible that we'll want to extend or replace them in the future, but
+currently the [`tracing-opentelemetry`](https://docs.rs/tracing-opentelemetry/)
+and [`opentelemetry-datadog`](https://docs.rs/opentelemetry-datadog) crates
+combine to fulfill this function to a satisfactory degree.
+
+## Rationale
+
+### Why is this change worth it?
+
+With a relatively small investment of effort, the system described here would be
+able to deliver significant, Vector-specific insight into performance directly
+from deployed environments. The type of instrumentation needed is already
+present within Vector, just underutilized, and it very neatly fills a gap
+between high-level metrics and low-level CPU profiling.
```

Comment:
Is there more of this gap we can fill without tracing? IIRC @jszwedko is tracking where we could have more metrics present in the system and I have thoughts as well from the performance work. There's a tension between the need for better observability into vector's operation and its throughput performance. I guess I want to be real sure we've wrung out all we can from cheaper methods before we start adding in more work that regresses throughput. 

I bring this up because the more I see tracing in our profiling output without getting real benefit from it the more tempted I am to ax it from the codebase.

Reply:
@blt 

> the more I see tracing in our profiling output without getting real benefit from it the more tempted I am to ax it from the codebase.

Could you be a little more specific here? The unfortunate collision between "tracing" the concept (discussed in this RFC, involving adding more spans and plumbing to ship them) and `tracing` the crate (as would show in current profiling output, used for logging and metrics context) makes this a little hard to parse. 

Axing the idea of shipping span data would be very simple, but axing `tracing` would require finding replacements for both internal logging and tagging of internal metrics. I've been a reasonably happy user of `tracing` so far, but if we have specific performance concerns around our current use then those should be resolved prior to further investment of the kind discussed in this RFC.

Reply:
Oh yeah, span data. I guess it is kind of an overloaded thing in our project. :) 

For instance, [this change](https://github.com/timberio/vector/pull/8579) ended up being worth +9Mb/s. The question for every span is then can we avoid its overhead or is the span providing information that's worth the overhead? We can maybe bound this by removing all spans and measuring vector's performance span-less.

Reply:
Gotcha, that makes sense. We have had at least one other major performance degradation due to unleveled spans, so it's definitely a concern.